extensions/sql/asset/loader/src/main/java/org/eclipse/dataspaceconnector/sql/asset/loader/SqlAssetLoader.java

```diff
@@ -0,0 +1,307 @@
+/*
+ *  Copyright (c) 2022 Daimler TSS GmbH
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Daimler TSS GmbH - Initial API and Implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.sql.asset.loader;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.eclipse.dataspaceconnector.dataloading.AssetEntry;
+import org.eclipse.dataspaceconnector.dataloading.AssetLoader;
+import org.eclipse.dataspaceconnector.spi.asset.AssetIndex;
+import org.eclipse.dataspaceconnector.spi.asset.AssetSelectorExpression;
+import org.eclipse.dataspaceconnector.spi.asset.DataAddressResolver;
+import org.eclipse.dataspaceconnector.spi.persistence.EdcPersistenceException;
+import org.eclipse.dataspaceconnector.spi.query.QuerySpec;
+import org.eclipse.dataspaceconnector.spi.transaction.TransactionContext;
+import org.eclipse.dataspaceconnector.spi.transaction.datasource.DataSourceRegistry;
+import org.eclipse.dataspaceconnector.spi.types.domain.DataAddress;
+import org.eclipse.dataspaceconnector.spi.types.domain.asset.Asset;
+import org.jetbrains.annotations.Nullable;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.AbstractMap;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.sql.DataSource;
+
+import static org.eclipse.dataspaceconnector.sql.SqlQueryExecutor.executeQuery;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.ASSET_COLUMN_ID;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.ASSET_PROPERTY_COLUMN_NAME;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.ASSET_PROPERTY_COLUMN_TYPE;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.ASSET_PROPERTY_COLUMN_VALUE;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.ASSET_PROPERTY_TABLE;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.ASSET_TABLE;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.DATA_ADDRESS_COLUMN_PROPERTIES;
+import static org.eclipse.dataspaceconnector.sql.asset.loader.SqlAssetTables.DATA_ADDRESS_TABLE;
+
+public class SqlAssetLoader implements AssetLoader, AssetIndex, DataAddressResolver {
```

Comment:
would rename to `SqlAssetIndex` as this is still the primary function of this class.